universum/analyzers/uncrustify.py

```diff
@@ -1,200 +1,202 @@
 import argparse
 import difflib
 import sys
-import os
+from os import environ
+from pathlib import Path
 
 import re
-import sh
+from typing import List, Optional, Tuple
 
 from . import utils
 
-# The maximum number of lines to write separate comments for
-# If exceeded, summarized comment will be provided instead
-MAX_LINES = 11
 
+class HtmlDiffFileWriter:
+
+    def __init__(self, target_folder: Path, wrapcolumn: int, tabsize: int) -> None:
+        self.differ = difflib.HtmlDiff(tabsize=tabsize, wrapcolumn=wrapcolumn)
+        self.target_folder = target_folder
+
+    def write_file(self, file_name: Path, src: List[str], target: List[str]) -> None:
+        relative_path = file_name.relative_to(Path.cwd())
+        out_file_name: str = str(relative_path).replace('/', '_') + '.html'
+        with open(self.target_folder.joinpath(out_file_name), 'w') as out_file:
+            out_file.write(self.differ.make_file(src, target, context=False))
+
+
+def form_arguments_for_documentation() -> argparse.ArgumentParser:
+    return _uncrustify_argument_parser()
+
+
+def main() -> int:
+    settings: argparse.Namespace = _uncrustify_argument_parser().parse_args()
+    src_folder: Path = Path.cwd()
+    target_folder: Path = Path.cwd().joinpath(settings.output_directory)
+    if not settings.cfg_file and 'UNCRUSTIFY_CONFIG' not in environ:
+        sys.stderr.write("Please specify the '--cfg_file' parameter "
+                         "or set an env. variable 'UNCRUSTIFY_CONFIG'")
+        return 2
+    wrapcolumn, tabsize = _get_wrapcolumn_tabsize(settings.cfg_file)
+    html_diff_file_writer = HtmlDiffFileWriter(target_folder, wrapcolumn, tabsize)
+
+    files: List[Tuple[Path, Path]] = []
+    try:
+        src_files: List[Path] = _get_src_files(settings.file_list, settings.file_list_config)
+        for pattern in settings.pattern_form:
+            regexp = re.compile(pattern)  # TODO: combine patterns via join
+            src_files = [file for file in src_files if regexp.match(str(file))]
+        for src_file in src_files:
+            rel_name = src_file.relative_to(src_folder)
+            target_file: Path = target_folder.joinpath(rel_name)
+            files.append((src_file, target_file))
+        if not files:
+            raise EnvironmentError("Please provide at least one file for analysis")
+    except EnvironmentError as e:
+        sys.stderr.write(str(e))
+        return 2
+
+    cmd = ["uncrustify", "-q", "-c", settings.cfg_file, "--prefix", settings.output_directory]
+    cmd.extend([str(path.relative_to(Path.cwd())) for path in src_files])
+
+    return utils.report_parsed_outcome(cmd,
+                                       lambda _: _uncrustify_output_parser(files, html_diff_file_writer, _),
```

Comment:
lambda is used to pass additional settings-based configuration data without modifying the function signature in 'utils' and other analyzers